### ISSUE TITLE
speed considerations for unrolling nodes and edges in the batch dimension?

### ISSUE THREAD

**Issue by pstjohn** [deepmind/graph_nets#16]
Most other libraries that have implemented message-passing neural networks (or other graph-network style approaches), pad the input graphs in batches according to something like
`node_features: (batch_size, max_nodes, node_features)`
`edge_features: (batch_size, max_edges, edge_features)`

This mirrors the typical approach in recurrent neural networks, where inputs of variable length are zero-padded to give a consistent batch size dimension.

In this library, the batch size is overloaded with the variable node and edge lengths; s.t. the inputs would be 
`node_features: (total_nodes_in_batch, node_features)`
`edge_features: (total_edges_in_batch, edge_features)`
where the total nodes and edges in each individual batch are subject to change. Are there any implementation-specific performance considerations to having the batch size changing on each iteration? I.e., are these models able to be effectively processed on GPUs, or is there a performance penalty to having to reallocate memory each gradient update?

On a related note, when I look at the device mapping of some of these models on GPUs, I notice that `utils_tf.repeat` typically gets allocated to the CPU, since there isn't a `cumsum` gpu optimization in tensorflow. Might be related to tensorflow/tensorflow#19570?

**Comment by pstjohn** [comment]
So I've rewritten some sections, but I'm seeing the GPU versions being maybe 3x as fast the CPU version.

GPU: 1/2 of a Tesla K80
CPU: Intel(R) Xeon(R) CPU E5-2640 v4 @ 2.40GHz

**Comment by vbapst** [comment]
We have benchmarks showing an (usual) 10x factor when running graph net code on GPU vs CPU, for a standard architecture, and this has also been our usual experience using this code.

This does not cover, however, the interesting case of variable batch size vs padding, but we will aim to provide more benchmark for this one in the coming weeks.

Also, just a note that one has to be cautious when benchmarking TF code to correctly take out all of the runtime outside the actual module execution (for instance, feeding data into the graph).

**Comment by ICDI0906** [comment]
> Most other libraries that have implemented message-passing neural networks (or other graph-network style approaches), pad the input graphs in batches according to something like
> `node_features: (batch_size, max_nodes, node_features)`
> `edge_features: (batch_size, max_edges, edge_features)`
> 
> This mirrors the typical approach in recurrent neural networks, where inputs of variable length are zero-padded to give a consistent batch size dimension.
> 
> In this library, the batch size is overloaded with the variable node and edge lengths; s.t. the inputs would be
> `node_features: (total_nodes_in_batch, node_features)`
> `edge_features: (total_edges_in_batch, edge_features)`
> where the total nodes and edges in each individual batch are subject to change. Are there any implementation-specific performance considerations to having the batch size changing on each iteration? I.e., are these models able to be effectively processed on GPUs, or is there a performance penalty to having to reallocate memory each gradient update?
> 
> On a related note, when I look at the device mapping of some of these models on GPUs, I notice that `utils_tf.repeat` typically gets allocated to the CPU, since there isn't a `cumsum` gpu optimization in tensorflow. Might be related to [tensorflow/tensorflow#19570](https://github.com/tensorflow/tensorflow/issues/19570)?

Hi~, I have the same question. Here I have X, which has a shape of [batch_size, timestep, node_size, node_feature_size], but I can't iterate on the timestep, because when using utils_tf.placeholders_from_networkxs function, it becomes [None,node_feature_size],Can you help me out? Looking forward to your reply!

**Comment by alvarosg** [comment]
When you use the placeholders, the key is to consider that the time axis is just another dimension of the features. This means that any sequence of graphs needs to have the same structure. This mean that your nodes would have shapes `[total_number_nodes_in_batch, sequence_length, node_feature_size]`.

**Comment by ICDI0906** [comment]
> When you use the placeholders, the key is to consider that the time axis is just another dimension of the features. This means that any sequence of graphs needs to have the same structure. This mean that your nodes would have shapes `[total_number_nodes_in_batch, sequence_length, node_feature_size]`.

Thanks, I use the shape as you said. And I split it on the sequence_lenght dimension, then feed it to GNN for every step's message passing!

**Comment by alvarosg** [comment]
(Closing due to inactivity)
